subprojects/core/src/main/java/org/gradle/api/internal/provider/AbstractLockableProperty.java

```diff
@@ -0,0 +1,62 @@
+/*
+ * Copyright 2017 the original author or authors.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.gradle.api.internal.provider;
+
+import javax.annotation.Nullable;
+
+public abstract class AbstractLockableProperty<T> extends AbstractProvider<T> implements PropertyInternal<T> {
+    private PropertyInternal<T> delegate;
+    private boolean locked;
+    private T value;
+
+    public AbstractLockableProperty(PropertyInternal<T> delegate) {
+        this.delegate = delegate;
+    }
+
+    @Override
+    public void setFromAnyValue(Object object) {
+        assertNotLocked();
+        delegate.setFromAnyValue(object);
+    }
+
+    @Nullable
+    @Override
+    public Class<T> getType() {
+        return delegate.getType();
```

Comment:
this will NPE after locking the property.

Reply:
Good catch, I fixed it and added a test for this case.